### PR TITLE
overlay tech debt: flow control

### LINKED ISSUE #3676: overlay tech debt: flow control
Since we've been adding new features to overlay, we've accumulated quite a bit of tech debt that would be useful to address before we add new features. I think this is worth investing resources into as it's becoming harder to unit test and validate new features in overlay because we don't have proper abstractions and separation of responsibilities. Auditing the code, we have an opportunity to clean up flow control functionality:
* Right now it's tightly coupled with the Peer class, making it hard to maintain and test. Ideally, we could extract flow control into its own module on top of peer. This way, Peer is solely responsible for the actual transport, authentication and forwarding network traffic to appropriate modules, while flow control ensures that we stop and resume reading/writing based on the throttling policy. 
* We should be able to unit test a narrow set of flow control policies without worrying about the actual internals of the Peer class and running the whole connection flow. This way we also don't need to duplicate test checks (e.g. peer must be authenticated, drop peer that doesn't respect the policy, etc)

### PR DESCRIPTION
**Pull request by marta-lokhova**
Resolves #3676 

First pass at decoupling Peer class into separate modules to ease code reviewing and testing. For now, I kept the tests as-is, but in the next iteration it'd be nice to audit and cleanup tests as well (maybe when we add bytes accounting). Specifically, it should now be easier to test flow-control-specific functionality (capacity accounting, outbound queuing, outbound purging etc) in an isolated way without depending on the internals of Peer/Application.

### CODE REVIEW

**Review by mbsdf — COMMENTED** [review]
LGTM so far

File: src/overlay/FlowControlCapacity.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  public:
+    virtual uint64_t getMsgResourceCount(StellarMessage const& msg) = 0;
+    virtual ReadingCapacity getCapacityLimits() const = 0;
+    virtual void releaseOutboundCapacity(StellarMessage const& msg) = 0;
+
+    void lockOutboundCapacity(StellarMessage const& msg);
+    bool lockLocalCapacity(StellarMessage const& msg);
+    uint64_t releaseLocalCapacity(StellarMessage const& msg);
+
+    bool hasOutboundCapacity(StellarMessage const& msg);
```
Comment (SirTyson) [inline src/overlay/FlowControlCapacity.h:41]: Const police

**Review by SirTyson — APPROVED** [review]
LGTM with a few typos and minor cleanups.

**Review by graydon — APPROVED** [review]
Quick skim (and sense that this is a no-op / refactor) LGTM

File: src/overlay/FlowControl.cpp [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    if (dropped && Logging::logTrace("Overlay"))
+    {
+        CLOG_TRACE(Overlay, "Dropped {} {} messages to peer {}", dropped,
+                   xdr::xdr_traits<MessageType>::enum_name(type),
+                   mApp.getConfig().toShortString(mNodeID));
+    }
+}
+
+Json::Value
+FlowControl::getFlowControlJsonInfo(bool compact)
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/overlay/FlowControl.cpp]: Done (other functions too)

**Review by mbsdf — APPROVED** [review]

**Comment by graydon** [comment]
r+ 11fcc58

**Review by mbsdf — COMMENTED** [review]
flow control in bytes PR #3737 is after this  overlay refactoring PR #3686  -  previous flow control in bytes PR #3607 is closed

**Pull request merged**